Fixing clipping

zvallisVividQ opened the pull request:
plus other changes:
- separating renderer
- more unit tests
- detection if cursor goes outside program window

zvallisVividQ committed:
correcting line clipping with 'out of window', mostly on right side

zvallisVividQ committed:
correcting line clipping with 'out of window', mostly on right side

zvallisVividQ committed:
correcting line clipping on entry to DW

zvallisVividQ committed:
correcting line clipping on entry to DW

zvallisVividQ committed:
tidying + release mode prepared

zvallisVividQ committed:
tidying + release mode prepared

lordGoldemort commented on PaintTests/WinTests.cpp:
```diff
@@ -609,7 +609,7 @@ namespace PaintTests
 			Renderer* rendererPtr = &dummyRenderer;
 			uint8_t ColA = 0;
```
Should be camelCase, not PascalCase
**SHOULD**

zvallisVividQ replied:
done

lordGoldemort commented on gfx/GFX_Renderer.h:
```diff
@@ -36,9 +36,9 @@ namespace gfx
 		void renderPresent() const;
```
Preferred structure to keep gfx free from SDL contamination

project gfx
```
    class Renderer {
        virtual void renderBox(..) = 0;
        virtual void renderText(…) = 0;
        // etc

    protected:
        Renderer() = default;
    };

```
project paint 
```
    class SDLRenderer : public gfx::Renderer {
    public:
        SDLRenderer(SDL_Renderer * sdl)
        : sdl_(sdl)
       {
       }
       
       virtual void renderBox(...) override;
       virtual void renderBox(...) override;
       virtual void renderBox(...) override;

    private:
        SDL_Renderer * sdlRenderer_;
    };
    

```
**SHOULD**

zvallisVividQ replied:
done

lordGoldemort commented on paint/PAINT_DrawWindow.cpp:
```diff
@@ -192,29 +204,24 @@ void DrawWindow::clearWindow() const
 //using Cohen-Sutherland clipping algorithm
 std::vector<win::Coords> DrawWindow::clippingHandler(win::Coords pStart, win::Coords pEnd) const
```
Prefer this to be a utility function in gfx

```
namespace gfx {
    std::vector<Point> clipLineToRect(Rect const & r, Point const & start, Point const & end);
}
```
            
(Would need to move Coord from win -> gfx OR create a Point type in gfx and a function to convert (hopefully just a cast!))
**SHOULD**

zvallisVividQ replied:
done, and moved Coord to gfx too

lordGoldemort commented on paint/PAINT_DrawWindow.h:
```diff
@@ -38,9 +38,10 @@ namespace paint
 		DrawWindow& operator=(const DrawWindow& that) = delete;
 		DrawWindow& operator=(DrawWindow&& that) = delete;
 
+		bool mouseEnter(bool clicked = false) override;
```
Just a comment. In future we might have the possibility of using the right mouse button to draw background colours whilst the left draws foreground. So a boolean woulndn't suffice here.

lordGoldemort commented on paint/PAINT_Eraser.h:
```diff
@@ -3,7 +3,7 @@
 namespace paint
 {
-	class Eraser final :
+	class Eraser :
```
Why remove `final`?

zvallisVividQ replied:
Resharper complained, I've added it back

lordGoldemort commented on paint/PAINT_StatusBarWindow.h:
```diff
@@ -9,7 +9,7 @@ namespace paint
 	public:
 		StatusBarWindow() = delete;
-		StatusBarWindow(gfx::Renderer* renderer, gfx::Rectangle const& rect, const char* name, const int drawWinX, const int drawWinY);
+		StatusBarWindow(gfx::Renderer* renderer, gfx::Rectangle const& rect, const char* name, int drawWinX, int drawWinY);
```
Don't construct with the renderer, just pass it into draw. Then there's no scope for doing renderer-related things in non-renderer-related contexts.
**MUST**

zvallisVividQ replied:
removed renderer_ pointer for windows where it's not necessary, and added draw overload with renderer

lordGoldemort commented on win/WIN_Slider.cpp:
```diff
@@ -183,7 +184,7 @@ bool Slider::mouseButtonUp(MouseButton button)
 		int yMouse = 0;
 		
-		SDL_GetMouseState(&xMouse, &yMouse);
+		getRenderer()->getMouseState(xMouse, yMouse);
```
Should probably just pass these in as parameters. This violates SRP (Single Responsibility Principle) - a Renderer should just be for rendering, not for getting peripheral state).
**MUST**

zvallisVividQ replied:
I put the mouse state function here to separate SDL from Slider (and other classes that rely on mouse states), but there's nothing that requires it be in Renderer. So I've moved getMouseState out of Renderer into win::utils.

lordGoldemort commented on win/WIN_Utils.h:
```diff
@@ -8,7 +8,7 @@ namespace win
-		bool filterNumerical(const char c);
+		bool filterNumerical(char c);
 
 		int findOutcode(gfx::Rectangle rect, int x, int y);
```
Should be part of gfx.
**MUST**

zvallisVividQ replied:
Done, outcode stuff is now in gfx. Win utils contains only filterNumerical and getMouseState now

lordGoldemort reviewed (approved):
ReSharper reports:

Solution paint.sln
    Project paint
      paint\PAINT_ButtonFunctions.cpp:18 Typo in comment: "cppcoreguidelines"
      paint\PAINT_ButtonFunctions.cpp:33 Typo in comment: "cppcoreguidelines"
      paint\PAINT_ButtonFunctions.cpp:78 Typo in comment: "Rerender"
      paint\PAINT_ColourDisplay.cpp:102 Typo in identifier: "cpick"
      paint\PAINT_ColourPicker.cpp:175 Typo in identifier: "Rerender"
      paint\PAINT_ColourPicker.h:38 Typo in identifier: "Rerender"
      paint\PAINT_DrawWindow.cpp:28 Typo in comment: "PIXELFORMAT"
      paint\PAINT_DrawWindow.cpp:28 Typo in comment: "TEXTUREACCESS"
      paint\PAINT_DrawWindow.cpp:169 Typo in comment: "cout"
      paint\PAINT_DrawWindow.cpp:171 Typo in comment: "cout"
      paint\PAINT_DrawWindow.cpp:192 Typo in identifier: "Rerender"
      paint\PAINT_DrawWindow.h:68 Typo in identifier: "Rerender"
      paint\PAINT_Program.cpp:91 Typo in comment: "rerender"
      paint\PAINT_Program.cpp:92 Typo in identifier: "rerender"
      paint\PAINT_Screen.cpp:121 Typo in identifier: "Rerender"
      paint\PAINT_Screen.h:32 Typo in identifier: "Rerender"
      paint\PAINT_Utils.cpp:36 Typo in comment: "cppcoreguidelines"

zvallisVividQ committed:
Retarget render functionality

please don't delete the retargeting renderer back to SCREEN in drawTool, that breaks everything and cannot figure out why to fix

zvallisVividQ committed:
pull request changes

zvallisVividQ committed:
Merge remote-tracking branch 'origin/zoe/fixing_dw_clipping' into fixing_dw_clipping

zvallisVividQ committed:
remove printf
